### PR TITLE
Implement notification endpoint

### LINKED ISSUE #2677: Implement notification changes
## Is your feature request related to a problem? Please describe 👀
Following on from #2459 - the notification query needs to be made available and then used by the front end.
From the epic:

* _Display excursions in the notification area, along with the most recent breach. An excursion disappears when the temperature is within the operating range_
* _Display the number of breaches and excursions in the notification area_

### PR DESCRIPTION
**Pull request by mark-prins**
Fixes #2677

# 👩🏻‍💻 What does this PR do? 
 
* Implements the graphQL query for excursions
* Changes the shape of the breaches part of the query ( to allow for total number of breaches to be shown )
* adds excursions into the notification area
* display the total number of breaches / excursions ( if there is more than one )

Screenshot 2024-01-09 at 12 12 27 PM

# 🧪 How has/should this change been tested? 

You will need some cold chain data. I suggest using the datafile in #2460 if you do not have one.
The excursion query will by default show you excursions from the last seven days ( the assumption is that old excursions are meaningless, and the query will be slow if we search across all temperature logs ). You may have to add an excursion manually - just create or edit a temperature log so that:

* the temperature is outside of the threshold ( check temperature_breach_config table ) with the default being 2-8°
* the date is within the past seven days

You should then see a notification appear. If you update the log entry, or add a later log entry that is within the temperature range and refresh the page, the notification will disappear.

Try adding an excursion for a different sensor in the same store - you'll see a message saying that there are now 2 excursions. 

## 📃 Documentation

_No user facing changes_

or

- [ ] Functional change 1
- [ ] Functional change 2

### CODE REVIEW

**Review by clemens-msupply — COMMENTED** [review]
Looks works fine!

I only have a question about the new breach config migration and the reason behind it. For example, it only inserts a config for existing stores in the DB but this doesn't work when it is a fresh initialization. This could lead to inconsitencies depending on how you setup your system, right?

how does the config syncs? (couldn't find it) will it be central data eventually?

Have the feeling its better to exclude the breach config migration, but might be wrong...

File: server/graphql/types/src/types/temperature_notification.rs [hunk 1]
```diff
@@ -69,11 +67,8 @@ impl TemperatureExcursionNode {
             .map(LocationNode::from_domain))
     }
 
-    pub async fn max_or_min_temperature(&self, ctx: &Context<'_>) -> Result<Option<f64>> {
-        Ok(get_max_or_min_breach_temperature(
-            &ctx.get_connection_manager().connection()?,
-            &self.row().id,
-        )?)
+    pub async fn max_or_min_temperature(&self) -> &f64 {
+        &self.row().temperature
```
[the remark this answers is not in the document]
Comment (mark-prins) [reply server/graphql/types/src/types/temperature_notification.rs:71]: 😄  I was thinking the same and had it as `temperature`! then changed it when implementing the front end; it is easier if the two types (breach & excursion) have the same shape. This probably is an indicator that I'm trying to do too much with the same objects.. but I've done it anyway

File: server/repository/src/migrations/v1_06_00/temperature_excursion.rs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+    sql!(
+        connection,
+        r#"
+            INSERT INTO temperature_breach_config (id, duration_milliseconds, "type", description, is_active, store_id, minimum_temperature, maximum_temperature)
+            SELECT '{id}', 0, 'EXCURSION', 'Default temperature excursion configuration', true, sp.id as store_id, 2, 8
+            FROM store_preference sp 
+            LEFT JOIN temperature_breach_config tbc ON sp.id = tbc.store_id AND tbc."type" = 'EXCURSION'
+            WHERE vaccine_module AND tbc.id is null;
+        "#,
```
[the remark this answers is not in the document]
Comment (mark-prins) [reply server/repository/src/migrations/v1_06_00/temperature_excursion.rs:15]: yeah, good point. I was thinking that we can add a default, and then have some way of editing that in the settings page. Given that the default won't be added as part of a new install, due to the missing store preferences, and won't be added if you change the store preference.. it's not a valid approach.

**Comment by mark-prins** [comment]
removed the migration code - it does mean that you will have to add a configuration record manually in order to see notifications. Will address that next.

**Review by clemens-msupply — APPROVED** [review]

**Pull request merged**